param/parameterized.py

```diff
@@ -1496,6 +1562,57 @@ def _instantiate_param(self_, param_obj, dict_=None, key=None):
             # should it instead keep the same name?
             new_object.param._generate_name()
 
+    def _update_deps(self_, attribute=None):
+        obj = self_.self
+        for n, queued, constant, deferred in type(obj).param._depends['watch']:
+            # On initialization set up constant watchers otherwise
+            # clean up previous dynamic watchers for the updated attribute
+            if attribute is None:
+                constant_grouped = defaultdict(list)
+                for dep in _resolve_mcs_deps(obj, constant, []):
+                    constant_grouped[(id(dep.inst), id(dep.cls), dep.what)].append((None, dep))
+                for group in constant_grouped.values():
+                    self_._watch_group(obj, n, queued, group)
+            else:
+                for w in obj._dynamic_watchers.get((n, attribute), []):
+                    (w.inst or w.cls).param.unwatch(w)
+
+            # Resolve dynamic dependencies one-by-one to be able to trace their watchers
+            grouped = defaultdict(list)
+            for ddep in deferred:
+                dattr = ddep.spec.split('.')[0]
+                if attribute and dattr != attribute:
+                    continue
+                for dep in _resolve_mcs_deps(obj, [], [ddep]):
+                    grouped[(id(dep.inst), id(dep.cls), dep.what, dattr)].append((ddep, dep))
+
+            for (_, _, _, dattr), group in grouped.items():
+                watcher = self_._watch_group(obj, n, queued, group)
+                obj._dynamic_watchers[(n, dattr)].append(watcher)
+
+    def _watch_group(self_, obj, name, queued, group):
+        ddeps = [dd for dd, _ in group if dd is not None]
+        _, gdep = group[0] # Need to grab representative dep from this group
```

Comment:
I wasn't able to trace through the logic to convince myself that group will always be non-empty. If it will,please put a comment explaining where that's ensured, so that [0] is safe to do.

Reply:
Wouldn't be much of a group if it didn't have any members now would it? 😉

Reply:
Groups here are basically the output of a `groupby` like operation so it makes no sense for them to be empty. Anyway the method now has a docstring.